5.0: Use RelationListIterator (#2548)
ezsystems/developer-documentation: docs/content_management/content_api/browsing_content.md

You can also use it to request other content-related value objects from various services:

> ``` php hl_lines="11"

Reviewer: ```suggestion
``` php hl_lines="13"
```

Author: Fixed in b324feb63d4081504ef8551602a5704ea9cfb76a

[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 1, 5) =]]

// …
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 24, 25) =]]
    // …
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 45, 46) =]]
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 53, 60) =]]
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 119) =]]
```

> `ContentInfo` is loaded from the [`ContentService`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentService.html) (line 8).

Reviewer: ```suggestion
`ContentInfo` is loaded from the [`ContentService`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentService.html).
```
Either line 13, or let's just remove it - it's hard to maintain.

Author: Fixed in b324feb63d4081504ef8551602a5704ea9cfb76a

[`URLAliasService::reverseLookup`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-URLAliasService.html#method_reverseLookup) gets the location's main [URL alias](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Values-Content-URLAlias.html):

``` php
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 63, 65, remove_indent=True) =]]
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 67, 69, remove_indent=True) =]]
```

### Content type

> You can retrieve the content type of content item through the [`getContentType`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Values-Content-ContentInfo.html#method_getContentType) method of the ContentInfo object:

Reviewer: ```suggestion
You can retrieve the content type of a content item through the [`getContentType`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Values-Content-ContentInfo.html#method_getContentType) method of the ContentInfo object:
```
I think the `a` is correct here?

### Relations

Content Relations are versioned.

> To list Relations to and from your content, you can to pass a `VersionInfo` object to the [`ContentService::loadRelationList`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentService.html#method_loadRelationList) method which is paginated.
> Or you can use the [`RelationListIteratorAdapter`](../../api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIteratorAdapter-RelationListIteratorAdapter.html) within a [`BatchIterator`](../../api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIterator.html).
> This method loads only the specified subset of relations to improve performance and was created with pagination in mind.

Reviewer: ```suggestion
To list Relations to and from your content, you can to pass a `VersionInfo` object to the [`ContentService::loadRelationList`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentService.html#method_loadRelationList) method.
This method loads only the specified subset of relations to improve performance and was created with pagination in mind.

Alternatively, use the [`RelationListIteratorAdapter`](../../api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIteratorAdapter-RelationListIteratorAdapter.html) within a [`BatchIterator`](../../api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIterator.html) to iterate over all relations.
```

I suggest to rewrite this part somehow.

First we mention the `loadRelationList`,then we mention `RelationListIteratorAdapter` object, only to start the next sentence with "This method` - so we're back to discussing loadRelationList?

Author: @mnocon Thank you for this highlight. That effectively doesn't make much sense. With the "you can to pass" grammar issue and this sentence about "pagination in mind" that doesn't sound like something I would write, there is clearly an issue here.

I rewrite this in d025c0cb4e69a013c9adb66cec2314b62866dd34
https://ez-systems-developer-documentation--2548.com.readthedocs.build/en/2548/content_management/content_api/browsing_content/#relations

### Relations

Content Relations are versioned.
To list Relations to and from your content, you can:

- pass a `VersionInfo` object to the [`ContentService::loadRelationList` method](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentService.html#method_loadRelationList) which returns a slice of the relation list thanks to pagination arguments
- use the [`RelationListIteratorAdapter`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIteratorAdapter-RelationListIteratorAdapter.html)

>   within a [`BatchIterator`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIterator.html) which allow traversing the relation list using one same object

Reviewer: ```suggestion
  within a [`BatchIterator`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Iterator-BatchIterator.html) which allows traversing the whole relation list
  
See [Processing large result sets](search_api.md#processing-large-result-sets) for more information about the BatchIterator.
```

Could you please add the RelationListAdapter to the list of Adapters on https://doc.ibexa.co/en/latest/search/search_api/#processing-large-result-sets ?

Author: Done in 8cdb1ab2f0608f350fc80f6ce5f8ebf3e70784da

All object state groups can be retrieved through [`loadObjectStateGroups`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ObjectStateService.html#method_loadObjectStateGroups).

``` php

> [[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 114, 118, remove_indent=True) =]]

Reviewer: I think this one should include the one line above, to include:

```
        $stateGroups = $this->objectStateService->loadObjectStateGroups();
```

as well.

```suggestion
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentMetaDataCommand.php', 113, 118, remove_indent=True) =]]
```

Author: Was fixed in 19daab97b0d7cdb3ee47835073d00605d747a16e

```

## Viewing content with fields

> To retrieve the fields of the selected content item, you can use the following command:

Reviewer: I think the header and description are wrong, and make it sound as if you need 3 services to access the value of a content item.

For that, you'd just use https://doc.ibexa.co/en/5.0/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-Values-Content-Content.html#method_getFields and iterate on that.

It might be out of scope for this PR and might need a follow-up - up to you.

```suggestion
## Viewing field definitions of content types

To retrieve the content type's field definitions of a selected content item, you can use the following command:
```

``` php hl_lines="17-18 20-27"
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentCommand.php', 1, 7) =]]

// …
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentCommand.php', 18, 19) =]]
    // …
[[= include_code('code_samples/api/public_php_api/src/Command/ViewContentCommand.php', 36) =]]
```

Line 17 shows how [`ContentService::loadContent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentService.html#method_loadContent) loads the content item provided to the command.
Line 18 makes use of the [`ContentTypeService`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-ContentTypeService.html) to retrieve the content type of the requested item.

Lines 20-27 iterate over fields defined by the content type.

> For each field they print out its identifier, and then using [`FieldTypeService`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-FieldTypeService.html) retrieve the field's value and print it out to the console.

Reviewer: ```suggestion
For each field definition they print out its identifier, and then using [`FieldTypeService`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-Core-Repository-FieldTypeService.html) retrieve the field definition's value and print it out to the console.
```
